File: bench/sqlite-vs-postgres/bench.mjs
import { Database } from "bun:sqlite";
import postgres from "postgres";
import { rmSync } from "fs";

// === CONFIG ===
const SEED_ROWS = 5000;
const WARMUP = 500;
const MEASURE = 5000;
const CONCURRENCIES = [1, 10, 50];

const SQLITE_PATH = "/tmp/bench_plausible.sqlite3";
try { rmSync(SQLITE_PATH); } catch {}
try { rmSync(SQLITE_PATH + "-wal"); } catch {}
try { rmSync(SQLITE_PATH + "-shm"); } catch {}

const sql = postgres({
  host: "127.0.0.1",
  port: 5432,
  database: "bench_plausible",
  username: "ubuntu",
  password: "bench",
  max: 50,
  idle_timeout: 5,
});

// === SCHEMA - mirip Plausible OLTP (users, teams, sites, memberships) ===
const DDL = `
DROP TABLE IF EXISTS team_memberships;
DROP TABLE IF EXISTS sites;
DROP TABLE IF EXISTS users;
DROP TABLE IF EXISTS teams;

CREATE TABLE teams (
  id SERIAL PRIMARY KEY,
  identifier TEXT NOT NULL UNIQUE,
  name TEXT NOT NULL,
  inserted_at TIMESTAMP DEFAULT NOW(),
  updated_at TIMESTAMP DEFAULT NOW()
);
CREATE TABLE users (
  id SERIAL PRIMARY KEY,
  email TEXT NOT NULL UNIQUE,
  name TEXT NOT NULL,
  team_id INT REFERENCES teams(id),
  inserted_at TIMESTAMP DEFAULT NOW(),
  updated_at TIMESTAMP DEFAULT NOW()
);
CREATE TABLE sites (
  id SERIAL PRIMARY KEY,
  domain TEXT NOT NULL UNIQUE,
  timezone TEXT DEFAULT 'Etc/UTC',
  team_id INT REFERENCES teams(id),
  public BOOLEAN DEFAULT false,
  inserted_at TIMESTAMP DEFAULT NOW(),
  updated_at TIMESTAMP DEFAULT NOW()
);
CREATE TABLE team_memberships (
  id SERIAL PRIMARY KEY,
  team_id INT NOT NULL REFERENCES teams(id),
  user_id INT NOT NULL REFERENCES users(id),
  role TEXT NOT NULL DEFAULT 'viewer',
  UNIQUE(team_id, user_id)
);
CREATE INDEX idx_sites_team_id ON sites(team_id);
CREATE INDEX idx_users_team_id ON users(team_id);
CREATE INDEX idx_sites_domain ON sites(domain);
CREATE INDEX idx_users_email ON users(email);
`;

const DDL_SQLITE = `
PRAGMA journal_mode=WAL;
PRAGMA synchronous=NORMAL;
PRAGMA cache_size=-64000;
PRAGMA temp_store=MEMORY;
DROP TABLE IF EXISTS team_memberships;
DROP TABLE IF EXISTS sites;
DROP TABLE IF EXISTS users;
DROP TABLE IF EXISTS teams;
CREATE TABLE teams (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  identifier TEXT NOT NULL UNIQUE,
  name TEXT NOT NULL,
  inserted_at TEXT DEFAULT (datetime('now')),
  updated_at TEXT DEFAULT (datetime('now'))
);
CREATE TABLE users (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  email TEXT NOT NULL UNIQUE,
  name TEXT NOT NULL,
  team_id INTEGER REFERENCES teams(id),
  inserted_at TEXT DEFAULT (datetime('now')),
  updated_at TEXT DEFAULT (datetime('now'))
);
CREATE TABLE sites (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  domain TEXT NOT NULL UNIQUE,
  timezone TEXT DEFAULT 'Etc/UTC',
  team_id INTEGER REFERENCES teams(id),
  public INTEGER DEFAULT 0,
  inserted_at TEXT DEFAULT (datetime('now')),
  updated_at TEXT DEFAULT (datetime('now'))
);
CREATE TABLE team_memberships (
  id INTEGER PRIMARY KEY AUTOINCREMENT,
  team_id INTEGER NOT NULL REFERENCES teams(id),
  user_id INTEGER NOT NULL REFERENCES users(id),
  role TEXT NOT NULL DEFAULT 'viewer',
  UNIQUE(team_id, user_id)
);
CREATE INDEX idx_sites_team_id ON sites(team_id);
CREATE INDEX idx_users_team_id ON users(team_id);
CREATE INDEX idx_sites_domain ON sites(domain);
CREATE INDEX idx_users_email ON users(email);
`;

console.log("== Setup schemas ==");
await sql.unsafe(DDL);
const db = new Database(SQLITE_PATH);
db.exec(DDL_SQLITE);

console.log(`== Seeding ${SEED_ROWS} rows each ==`);
// Seed deterministically - batch insert for speed
// Postgres: use COPY-like batch
let teams = Array.from({ length: 500 }, (_, i) => ({
  identifier: `team-${i.toString().padStart(4, "0")}-xxxxxxxx`,
  name: `Team ${i}`,
}));

for (let t of teams) {
  await sql`INSERT INTO teams (identifier, name) VALUES (${t.identifier}, ${t.name})`;
}
const pgTeams = await sql`SELECT id FROM teams ORDER BY id`;
console.log(`  PG teams: ${pgTeams.length}`);

for (let i = 0; i < SEED_ROWS; i++) {
  const team_id = pgTeams[i % pgTeams.length].id;
  const email = `user${i}@example.com`;
  const domain = `site${i}.example.com`;
  await sql`INSERT INTO users (email, name, team_id) VALUES (${email}, ${`User ${i}`}, ${team_id})`;
  await sql`INSERT INTO sites (domain, team_id) VALUES (${domain}, ${team_id})`;
}
const pgUsers = await sql`SELECT id, team_id FROM users ORDER BY id`;
for (let i = 0; i < pgUsers.length; i++) {
  await sql`INSERT INTO team_memberships (team_id, user_id, role) VALUES (${pgUsers[i].team_id}, ${pgUsers[i].id}, 'viewer') ON CONFLICT DO NOTHING`;
}
console.log(`  PG users/sites: ${SEED_ROWS}`);

// Seed SQLite identically via prepared statements (fast)
db.exec("BEGIN");
const insTeam = db.prepare("INSERT INTO teams (identifier, name) VALUES (?, ?)");
const insUser = db.prepare("INSERT INTO users (email, name, team_id) VALUES (?, ?, ?)");
const insSite = db.prepare("INSERT INTO sites (domain, team_id) VALUES (?, ?)");
const insMem = db.prepare("INSERT OR IGNORE INTO team_memberships (team_id, user_id, role) VALUES (?, ?, 'viewer')");
for (let t of teams) insTeam.run(t.identifier, t.name);
const sqliteTeams = db.prepare("SELECT id FROM teams ORDER BY id").all();
for (let i = 0; i < SEED_ROWS; i++) {
  const team_id = sqliteTeams[i % sqliteTeams.length].id;
  insUser.run(`user${i}@example.com`, `User ${i}`, team_id);
  insSite.run(`site${i}.example.com`, team_id);
}
const sqliteUsers = db.prepare("SELECT id, team_id FROM users ORDER BY id").all();
for (let u of sqliteUsers) insMem.run(u.team_id, u.id);
db.exec("COMMIT");
console.log(`  SQLite users/sites: ${SEED_ROWS}`);
console.log(`  SQLite file: ${(Bun.file(SQLITE_PATH).size / 1024 / 1024).toFixed(2)} MB`);

// Prepare statements
const sqliteStmtSiteByDomain = db.prepare("SELECT * FROM sites WHERE domain = ? LIMIT 1");
const sqliteStmtUserByEmail = db.prepare("SELECT u.*, t.name as team_name FROM users u JOIN teams t ON t.id = u.team_id WHERE u.email = ? LIMIT 1");
const sqliteStmtSitesByTeam = db.prepare("SELECT * FROM sites WHERE team_id = ? ORDER BY domain LIMIT 20");
const sqliteStmtInsertSite = db.prepare("INSERT INTO sites (domain, team_id) VALUES (?, ?)");
const sqliteStmtUpdateSite = db.prepare("UPDATE sites SET timezone = ? WHERE id = ?");

// Postgres helpers
async function pgSiteByDomain(domain) { return sql`SELECT * FROM sites WHERE domain = ${domain} LIMIT 1`; }
async function pgUserByEmail(email) { return sql`SELECT u.*, t.name as team_name FROM users u JOIN teams t ON t.id = u.team_id WHERE u.email = ${email} LIMIT 1`; }
async function pgSitesByTeam(teamId) { return sql`SELECT * FROM sites WHERE team_id = ${teamId} ORDER BY domain LIMIT 20`; }
async function pgInsertSite(domain, teamId) { return sql`INSERT INTO sites (domain, team_id) VALUES (${domain}, ${teamId})`; }
async function pgUpdateSite(id, tz) { return sql`UPDATE sites SET timezone = ${tz} WHERE id = ${id}`; }

// === BENCH HARNESS ===
function percentile(arr, p) {
  const s = [...arr].sort((a, b) => a - b);
  return s[Math.floor(s.length * p)];
}

async function runBench(name, sqliteFn, pgFn, concurrency) {
  // warmup
  for (let i = 0; i < WARMUP; i++) {
    const k = i % SEED_ROWS;
    await sqliteFn(k);
    await pgFn(k);
  }

  // measure sqlite
  let sqliteLat = [];
  {
    const start = performance.now();
    // concurrency: batch promises (for sqlite, sequential per connection but we simulate via workers)
    // SQLite bun:sqlite is sync - measure sync loop vs async pg
    // For fair comparison, run sqlite in same async batch pattern but sync calls are fast
    let idx = 0;
    async function worker() {
      for (let i = 0; i < MEASURE / concurrency; i++) {
        const k = (idx++ % SEED_ROWS);
        const t0 = performance.now();
        sqliteFn(k);
        sqliteLat.push(performance.now() - t0);
      }
    }
    await Promise.all(Array.from({ length: concurrency }, () => worker()));
    var sqliteTime = performance.now() - start;
  }

  let pgLat = [];
  {
    const start = performance.now();
    let idx = 0;
    async function worker() {
      for (let i = 0; i < MEASURE / concurrency; i++) {
        const k = (idx++ % SEED_ROWS);
        const t0 = performance.now();
        await pgFn(k);
        pgLat.push(performance.now() - t0);
      }
    }
    await Promise.all(Array.from({ length: concurrency }, () => worker()));
    var pgTime = performance.now() - start;
  }

  const sqliteQps = (MEASURE / (sqliteTime / 1000));
  const pgQps = (MEASURE / (pgTime / 1000));
  return {
    name, concurrency,
    sqliteQps: Math.round(sqliteQps), pgQps: Math.round(pgQps),
    sqliteP50: percentile(sqliteLat, 0.5).toFixed(3),
    sqliteP95: percentile(sqliteLat, 0.95).toFixed(3),
    sqliteP99: percentile(sqliteLat, 0.99).toFixed(3),
    pgP50: percentile(pgLat, 0.5).toFixed(3),
    pgP95: percentile(pgLat, 0.95).toFixed(3),
    pgP99: percentile(pgLat, 0.99).toFixed(3),
    speedup: (sqliteQps / pgQps).toFixed(2),
  };
}

console.log("\n== Running benchmarks (MEASURE=" + MEASURE + ", WARMUP=" + WARMUP + ") ==");

// Define workloads - representative of Plausible single-server traffic
const workloads = [
  {
    name: "Q1: site by domain (point lookup, 70% traffic)",
    sqliteFn: (k) => sqliteStmtSiteByDomain.get(`site${k}.example.com`),
    pgFn: (k) => pgSiteByDomain(`site${k}.example.com`),
  },
  {
    name: "Q2: user by email + JOIN team (auth, 15% traffic)",
    sqliteFn: (k) => sqliteStmtUserByEmail.get(`user${k}@example.com`),
    pgFn: (k) => pgUserByEmail(`user${k}@example.com`),
  },
  {
    name: "Q3: list sites by team (dashboard, 10% traffic)",
    sqliteFn: (k) => sqliteStmtSitesByTeam.all((k % 500) + 1),
    pgFn: (k) => pgSitesByTeam((k % 500) + 1),
  },
];

let results = [];
for (let w of workloads) {
  for (let c of CONCURRENCIES) {
    console.log(`  bench ${w.name} c=${c}...`);
    const r = await runBench(w.name, w.sqliteFn, w.pgFn, c);
    results.push(r);
    console.log(`    SQLite ${r.sqliteQps} qps p95=${r.sqliteP95}ms | PG ${r.pgQps} qps p95=${r.pgP95}ms | x${r.speedup}`);
  }
}

// Write-heavy workload separately (SQLite writer lock) - sequential to avoid SQLITE_BUSY noise
console.log("\n== Q4: mixed read 95% / write 5% (realistic) c=10 ==");
{
  let sqliteLat = [], pgLat = [];
  // warmup
  for (let i = 0; i < 200; i++) {
    sqliteStmtSiteByDomain.get(`site${i % SEED_ROWS}.example.com`);
    await pgSiteByDomain(`site${i % SEED_ROWS}.example.com`);
  }
  const N = 2000;
  // SQLite mixed
  let t0 = performance.now();
  for (let i = 0; i < N; i++) {
    const s = performance.now();
    if (i % 20 === 0) {
      const id = (i % SEED_ROWS) + 1;
      sqliteStmtUpdateSite.run("Asia/Jakarta", id);
    } else {
      sqliteStmtSiteByDomain.get(`site${(i*7)%SEED_ROWS}.example.com`);
    }
    sqliteLat.push(performance.now() - s);
  }
  const sqliteTime = performance.now() - t0;

  t0 = performance.now();
  for (let i = 0; i < N; i++) {
    const s = performance.now();
    if (i % 20 === 0) {
      const id = (i % SEED_ROWS) + 1;
      await pgUpdateSite(id, "Asia/Jakarta");
    } else {
      await pgSiteByDomain(`site${(i*7)%SEED_ROWS}.example.com`);
    }
    pgLat.push(performance.now() - s);
  }
  const pgTime = performance.now() - t0;
  const r = {
    name: "Q4: mixed 95% read / 5% write",
    sqliteQps: Math.round(N / (sqliteTime/1000)),
    pgQps: Math.round(N / (pgTime/1000)),
    sqliteP95: percentile(sqliteLat, 0.95).toFixed(3),
    pgP95: percentile(pgLat, 0.95).toFixed(3),
  };
  r.speedup = (r.sqliteQps / r.pgQps).toFixed(2);
  results.push({ ...r, concurrency: 1, sqliteP50: percentile(sqliteLat,0.5).toFixed(3), sqliteP99: percentile(sqliteLat,0.99).toFixed(3), pgP50: percentile(pgLat,0.5).toFixed(3), pgP99: percentile(pgLat,0.99).toFixed(3) });
  console.log(`  SQLite ${r.sqliteQps} qps p95=${r.sqliteP95}ms | PG ${r.pgQps} qps p95=${r.pgP95}ms | x${r.speedup}`);
}

// Resource
console.log("\n== Resource ==");
import { $ } from "bun";
const pgMem = await $`ps -o rss,comm -p $(pgrep -d, -x postgres) 2>&1 | awk 'NR>1{sum+=$1} END{print sum/1024 " MB"}'`.text();
console.log(`  Postgres RSS total: ${pgMem.trim()}`);
console.log(`  SQLite RSS extra: 0 MB (in-process, ~2-5 MB heap)`);
const sqliteSize = (Bun.file(SQLITE_PATH).size / 1024).toFixed(0);
console.log(`  SQLite file: ${sqliteSize} KB | PG data: $(du -sh /var/lib/postgresql/18/main | cut -f1)`);

console.log("\n== RESULTS JSON ==");
console.log(JSON.stringify(results, null, 2));

await sql.end();
db.close();

console.log("\nDone. SQLite WAL file exists:", Bun.file(SQLITE_PATH + "-wal").size);

File: bench/ecto-vs-postgres/bench.exs
Logger.configure(level: :warning)
import Ecto.Query

# Wait for repos
Process.sleep(500)

# ── Setup: recreate SQLite schema ──
Bench.SqliteRepo.query!("PRAGMA journal_mode=WAL;")
Bench.SqliteRepo.query!("PRAGMA synchronous=NORMAL;")
Bench.SqliteRepo.query!("PRAGMA cache_size=-64000;")
Bench.SqliteRepo.query!("PRAGMA temp_store=MEMORY;")

# Create tables if not exist (idempotent) - drop first for clean state
for sql <- [
  "DROP TABLE IF EXISTS team_memberships",
  "DROP TABLE IF EXISTS sites",
  "DROP TABLE IF EXISTS users",
  "DROP TABLE IF EXISTS teams",
  """
  CREATE TABLE teams (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    identifier TEXT NOT NULL UNIQUE,
    name TEXT NOT NULL,
    inserted_at TEXT DEFAULT (datetime('now')),
    updated_at TEXT DEFAULT (datetime('now'))
  )
  """,
  """
  CREATE TABLE users (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    email TEXT NOT NULL UNIQUE,
    name TEXT NOT NULL,
    team_id INTEGER REFERENCES teams(id),
    inserted_at TEXT DEFAULT (datetime('now')),
    updated_at TEXT DEFAULT (datetime('now'))
  )
  """,
  """
  CREATE TABLE sites (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    domain TEXT NOT NULL UNIQUE,
    timezone TEXT DEFAULT 'Etc/UTC',
    team_id INTEGER REFERENCES teams(id),
    inserted_at TEXT DEFAULT (datetime('now')),
    updated_at TEXT DEFAULT (datetime('now'))
  )
  """,
  """
  CREATE TABLE team_memberships (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    team_id INTEGER NOT NULL REFERENCES teams(id),
    user_id INTEGER NOT NULL REFERENCES users(id),
    role TEXT NOT NULL DEFAULT 'viewer',
    UNIQUE(team_id, user_id)
  )
  """,
  "CREATE INDEX IF NOT EXISTS idx_sites_team_id ON sites(team_id)",
  "CREATE INDEX IF NOT EXISTS idx_users_team_id ON users(team_id)",
  "CREATE INDEX IF NOT EXISTS idx_sites_domain ON sites(domain)",
  "CREATE INDEX IF NOT EXISTS idx_users_email ON users(email)"
] do
  Bench.SqliteRepo.query!(sql)
end

IO.puts("== Seeding SQLite 5000 rows ==")
Bench.SqliteRepo.query!("BEGIN")
for i <- 0..499 do
  Bench.SqliteRepo.query!("INSERT INTO teams (identifier, name) VALUES (?, ?)", ["team-#{String.pad_leading(to_string(i), 4, "0")}-xxxxxxxx", "Team #{i}"])
end
team_ids = Bench.SqliteRepo.query!("SELECT id FROM teams ORDER BY id") |> Map.get(:rows) |> Enum.map(&hd/1)
for i <- 0..4999 do
  team_id = Enum.at(team_ids, rem(i, length(team_ids)))
  Bench.SqliteRepo.query!("INSERT INTO users (email, name, team_id) VALUES (?, ?, ?)", ["user#{i}@example.com", "User #{i}", team_id])
  Bench.SqliteRepo.query!("INSERT INTO sites (domain, team_id) VALUES (?, ?)", ["site#{i}.example.com", team_id])
end
user_rows = Bench.SqliteRepo.query!("SELECT id, team_id FROM users ORDER BY id") |> Map.get(:rows)
for [uid, tid] <- user_rows do
  Bench.SqliteRepo.query!("INSERT OR IGNORE INTO team_memberships (team_id, user_id, role) VALUES (?, ?, 'viewer')", [tid, uid])
end
Bench.SqliteRepo.query!("COMMIT")
IO.puts("  SQLite seeded")

# Postgres is already seeded by Bun bench (5000 rows). Verify count
pg_count = Bench.PgRepo.one(from s in "sites", select: count(s.id))
IO.puts("  PG sites count: #{pg_count} (already seeded by Bun bench, reusing)")

# ── Define workload fns (Ecto) ──
defmodule Work do
  import Ecto.Query
  def sqlite_site_by_domain(k), do: Bench.SqliteRepo.one(from s in Bench.Site, where: s.domain == ^"site#{k}.example.com", limit: 1)
  def pg_site_by_domain(k), do: Bench.PgRepo.one(from s in Bench.Site, where: s.domain == ^"site#{k}.example.com", limit: 1)

  def sqlite_user_by_email(k), do: Bench.SqliteRepo.one(from u in Bench.User, join: t in assoc(u, :team), where: u.email == ^"user#{k}@example.com", select: %{email: u.email, team_name: t.name}, limit: 1)
  def pg_user_by_email(k), do: Bench.PgRepo.one(from u in Bench.User, join: t in assoc(u, :team), where: u.email == ^"user#{k}@example.com", select: %{email: u.email, team_name: t.name}, limit: 1)

  def sqlite_sites_by_team(k), do: Bench.SqliteRepo.all(from s in Bench.Site, where: s.team_id == ^(rem(k, 500)+1), order_by: s.domain, limit: 20)
  def pg_sites_by_team(k), do: Bench.PgRepo.all(from s in Bench.Site, where: s.team_id == ^(rem(k, 500)+1), order_by: s.domain, limit: 20)
end

# Warmup
IO.puts("\n== Warmup 500 ==")
for i <- 0..499 do
  k = rem(i, 5000)
  Work.sqlite_site_by_domain(k)
  Work.pg_site_by_domain(k)
end

# ── Benchee single-concurrency ──
IO.puts("\n== Benchee (single concurrency, Ecto overhead) ==")
Benchee.run(
  %{
    "sqlite site by domain" => fn -> Work.sqlite_site_by_domain(Enum.random(0..4999)) end,
    "pg site by domain" => fn -> Work.pg_site_by_domain(Enum.random(0..4999)) end,
    "sqlite user JOIN team" => fn -> Work.sqlite_user_by_email(Enum.random(0..4999)) end,
    "pg user JOIN team" => fn -> Work.pg_user_by_email(Enum.random(0..4999)) end,
    "sqlite list sites by team" => fn -> Work.sqlite_sites_by_team(Enum.random(0..499)) end,
    "pg list sites by team" => fn -> Work.pg_sites_by_team(Enum.random(0..499)) end
  },
  time: 5,
  warmup: 2,
  print: [benchmarking: true, configuration: false]
)

# ── Custom concurrency harness ──
defmodule Harness do
  def percentile(list, p) do
    sorted = Enum.sort(list)
    idx = trunc(length(sorted) * p)
    Enum.at(sorted, idx)
  end

  def run(name, sqlite_fn, pg_fn, concurrency, n \\ 5000) do
    # sqlite
    {sqlite_time, sqlite_lat} = :timer.tc(fn ->
      1..concurrency
      |> Task.async_stream(fn _ ->
        for i <- 0..div(n, concurrency)-1 do
          k = rem(i, 5000)
          t0 = System.monotonic_time(:microsecond)
          sqlite_fn.(k)
          System.monotonic_time(:microsecond) - t0
        end
      end, max_concurrency: concurrency, timeout: 30_000)
      |> Enum.flat_map(fn {:ok, v} -> v end)
    end)
    sqlite_qps = trunc(n / (sqlite_time / 1_000_000))
    sqlite_p50 = percentile(sqlite_lat, 0.5) / 1000
    sqlite_p95 = percentile(sqlite_lat, 0.95) / 1000
    sqlite_p99 = percentile(sqlite_lat, 0.99) / 1000

    # pg
    {pg_time, pg_lat} = :timer.tc(fn ->
      1..concurrency
      |> Task.async_stream(fn _ ->
        for i <- 0..div(n, concurrency)-1 do
          k = rem(i, 5000)
          t0 = System.monotonic_time(:microsecond)
          pg_fn.(k)
          System.monotonic_time(:microsecond) - t0
        end
      end, max_concurrency: concurrency, timeout: 30_000)
      |> Enum.flat_map(fn {:ok, v} -> v end)
    end)
    pg_qps = trunc(n / (pg_time / 1_000_000))
    pg_p50 = percentile(pg_lat, 0.5) / 1000
    pg_p95 = percentile(pg_lat, 0.95) / 1000
    pg_p99 = percentile(pg_lat, 0.99) / 1000

    speedup = sqlite_qps / max(pg_qps, 1)
    IO.puts("#{name} c=#{concurrency} | SQLite #{sqlite_qps} qps p95=#{Float.round(sqlite_p95, 3)}ms | PG #{pg_qps} qps p95=#{Float.round(pg_p95, 3)}ms | x#{Float.round(speedup, 2)}  (sqlite p50 #{Float.round(sqlite_p50,3)} p99 #{Float.round(sqlite_p99,3)} | pg p50 #{Float.round(pg_p50,3)} p99 #{Float.round(pg_p99,3)})")
    {sqlite_qps, pg_qps, speedup}
  end
end

IO.puts("\n== Concurrency sweep (Ecto, Task.async_stream) MEASURE=5000 ==")
for c <- [1, 10, 50] do
  Harness.run("Q1 site by domain", &Work.sqlite_site_by_domain/1, &Work.pg_site_by_domain/1, c)
end
for c <- [1, 10, 50] do
  Harness.run("Q2 user JOIN", &Work.sqlite_user_by_email/1, &Work.pg_user_by_email/1, c)
end
for c <- [1, 10, 50] do
  Harness.run("Q3 list by team", &Work.sqlite_sites_by_team/1, &Work.pg_sites_by_team/1, c)
end

# Mixed 95/5
IO.puts("\n== Q4 mixed 95% read 5% write (sequential, 2000 ops) ==")
sqlite_lat = for i <- 0..1999 do
  t0 = System.monotonic_time(:microsecond)
  if rem(i, 20) == 0 do
    Bench.SqliteRepo.query!("UPDATE sites SET timezone = ? WHERE id = ?", ["Asia/Jakarta", rem(i, 5000)+1])
  else
    Work.sqlite_site_by_domain(rem(i*7, 5000))
  end
  System.monotonic_time(:microsecond) - t0
end
pg_lat = for i <- 0..1999 do
  t0 = System.monotonic_time(:microsecond)
  if rem(i, 20) == 0 do
    Bench.PgRepo.query!("UPDATE sites SET timezone = $1 WHERE id = $2", ["Asia/Jakarta", rem(i, 5000)+1])
  else
    Work.pg_site_by_domain(rem(i*7, 5000))
  end
  System.monotonic_time(:microsecond) - t0
end
IO.puts("  SQLite p50 #{Harness.percentile(sqlite_lat,0.5)/1000}ms p95 #{Harness.percentile(sqlite_lat,0.95)/1000}ms p99 #{Harness.percentile(sqlite_lat,0.99)/1000}ms")
IO.puts("  PG     p50 #{Harness.percentile(pg_lat,0.5)/1000}ms p95 #{Harness.percentile(pg_lat,0.95)/1000}ms p99 #{Harness.percentile(pg_lat,0.99)/1000}ms")

IO.puts("\nDone")
